### ISSUE TITLE
/bin/sh: 1: raxmlHPC-AVX: not found

### ISSUE THREAD

**Issue by davidemms** [davidemms/OrthoFinder#266]
Copied message from [here](https://github.com/davidemms/OrthoFinder/commit/044dc91e400690b445c9736260f23b1c11285f5a#commitcomment-33797372) as a new issue:

dear Sir,

I Edited the Config.json file, but still i get error

/bin/sh: 1: raxmlHPC-AVX: not found

ERROR: Cannot run user-configured tree method 'raxml'
Please check program is installed and that it is correctly configured in the orthofinder/config.json file

Dependencies have been met for inference of orthogroups but not for the subsequent orthologue inference.
Either install the required dependencies or use the option '-og' to stop the analysis after the inference of orthogroups.

ERROR: An error occurred, please review error messages for more information.

**Comment by davidemms** [comment]
Hi

OrthoFinder is attempting to run an executable called "raxmlHPC-AVX". You should check that this executable is in the system path on your machine and that it runs correctly by attempting to run it from the the same command prompt you are using to run orthofinder. If your RAxML executable has a different name then you can change the name of the executable or change the "cmd_line" entry in the config file to resolve that.

All the best
David

**Comment by ens-LCampbell** [comment]
Hi David,

I do not feel this issue is resolved yet. I to have the exact same issue when trying to run Raxml.

Test can run "raxml" - failed

Traceback (most recent call last):
  File "/hps/nobackup2/production/ensemblgenomes/lcampbell/Software/OrthoFinder/orthofinder.py", line 7, in <module>
    main(args)
  File "/hps/nobackup2/production/ensemblgenomes/lcampbell/Software/OrthoFinder/scripts_of/__main__.py", line 1643, in main
    CheckDependencies(options, prog_caller, files.FileHandler.GetWorkingDirectory1_Read()[0])
  File "/hps/nobackup2/production/ensemblgenomes/lcampbell/Software/OrthoFinder/scripts_of/__main__.py", line 1303, in CheckDependencies
    options.qStopAfterAlignments):
  File "/hps/nobackup2/production/ensemblgenomes/lcampbell/Software/OrthoFinder/scripts_of/orthologues.py", line 622, in CanRunOrthologueDependencies
    if not program_caller.TestTreeMethod(temp_dir, tree_method):
  File "/hps/nobackup2/production/ensemblgenomes/lcampbell/Software/OrthoFinder/scripts_of/program_caller.py", line 180, in TestTreeMethod
    return self._TestMethod(working_dir, 'tree', method_name)
  File "/hps/nobackup2/production/ensemblgenomes/lcampbell/Software/OrthoFinder/scripts_of/program_caller.py", line 230, in _TestMethod
    print(("".join(stderr)))
TypeError: sequence item 0: expected str instance, bytes found

This issue first appeared when running on an LSF cluster, where my command line running of orthoFinder would run fine, raxml would be located. But when submitting the job via bsub to our cluster scheduler, raxml would then fail just like above. Something similar happens when I create an interactive shell environment (same as via Bsub but where you're logged onto the worker node as the job runs). Same story again, trying to run with raxml inference, it fails to locate any raxml installation. 

I am 100% sure that its installed as i can literally call it from the command line 

$raxmlHPC-AVX. 

if I call which raxmlHPS-AVX

returns: 
/nfs/software/ensembl/RHEL7-JUL2017-core2/linuxbrew/bin/raxmlHPC-AVX

I've tried adding the full Path, didn't work. I tried creating aliases didn't work. The config.json file is exactly the same as I call it on the command line but it won't work. I've added the executable to the bin dir "OrthoFinder/scripts_of/bin" of Orthofinder, that didn't work. 

Please can you advise here, tis quite a frustrating issue ha :) 

Thank you David
Lahcen

**Comment by davidemms** [comment]
Hi Lahcen

Thanks for reporting this. At the moment I don't know what is causing this for you, but I can see in the output you've posted that the error message from trying to call RAxML is not being reported properly:

> 
> print(("".join(stderr)))
> TypeError: sequence item 0: expected str instance, bytes found

I've just submitted a fix for that (adcf57fbf47c03556eedd6f16427949da4b4bca0) so that you should get the error message properly. That might reveal where the problem is or point to the next steps to try. Let me know how you get on.

All the best
David

**Comment by ens-LCampbell** [comment]
Hi David,

Many thanks for the speedy reply ! 

I've since pulled your changes and thank you it has helped print a more useful error statement. 

"_Test can run "mafft /hps/nobackup2/production/ensembl/lcampbell_ens-met_prod_1/OrthoFinder/compara_MLSS-sp_SSA1Whitefly_run1/OrthoFinder/Results_Apr08/WorkingDirectory//_dependencies_check/SimpleTest.fa" - ok
Test can run "raxml" - failed

/bin/sh: line 1: 99743 Illegal instruction     raxmlHPC-AVX -m PROTGAMMALG -p 12345 -s /hps/nobackup2/production/ensembl/lcampbell_ens-met_prod_1/OrthoFinder/compara_MLSS-sp_SSA1Whitefly_run1/OrthoFinder/Results_Apr08/WorkingDirectory//_dependencies_check/temp_83583209132/Species0.fa -n test -w /hps/nobackup2/production/ensembl/lcampbell_ens-met_prod_1/OrthoFinder/compara_MLSS-sp_SSA1Whitefly_run1/OrthoFinder/Results_Apr08/WorkingDirectory//_dependencies_check/temp_83583209132 > /dev/null

ERROR: Cannot run user-configured tree method 'raxml'
Please check program is installed and that it is correctly configured in the orthofinder/config.json file

Dependencies have been met for inference of orthogroups but not for the subsequent orthologue inference.
Either install the required dependencies or use the option '-og' to stop the analysis after the inference of orthogroups."_

I have contacted our system admins for any insight as Im not quite sure what the main source of the problem is. Perhaps partial loading of a user ENV when bsub'ing a job to run Orthofinder. 

On our head nodes it works fine, but when bsub'ed to a worker node most of the time it returns the no raxml error. I can let you know if systems here figure out why this is happening. 

Checking required programs are installed
----------------------------------------
Test can run "mafft /hps/nobackup2/production/ensembl/lcampbell_ens-met_prod_1/OrthoFinder/compara_MLSS-sp_SSA1Whitefly_run1/OrthoFinder/Results_Apr08/WorkingDirectory//_dependencies_check/SimpleTest.fa" - ok
Test can run "raxml" - ok

Quite perplexed here :) 

Best
Lahcen

**Comment by davidemms** [comment]
Ok, so from the error it looks like /bin/sh is producing the error message rather than raxml. Also, I think it's not a problem finding or running raxml but that instead it does not like the command. I.e. is says "Illegal instruction". 

My first guess: I wonder if perhaps it doesn't like the "> /dev/null" at the end. This is only there to suppress the output to screen from each of the commands--a single tree can produce pages of output so imagine what 20000 trees will do! However, I think the output might be suppressed by OrthoFinder higher up anyway.

I would try editing the config.json file to remove "> /dev/null" from the command. If it does produce a load of output to screen you could always look in the raxml help to see if there is a way to silence it.

If this isn't it then let me know and we think about what else it could be.

All the best
David

**Comment by ens-LCampbell** [comment]
Hi David,

So I've tried that and sadly all it does is remove the previous failure error. It no longer complains about bin/sh. But it doesn't say anything else, just that "raxml" - failed. 

I honestly feel its something more related to how our cluster is set up and if there are gremlins for specific nodes. As I say head/login nodes everything seems to run ok, all deps are located. Its when submitted the issues creep up. 

Can I ask, is there anyway in the future to have users simply transfer ALL the deps executable needed inside OrthoFinder/scripts_of/bin ? Not sure who much this would make things easier but could be an idea perhaps down the line. 

I will let you know if we find any bug on our end ! 
Lahcen

**Comment by davidemms** [comment]
Hi Lahcen

You can definitely do that provided your executables run on all your servers and you don't need different versions. Just put whatever binaries you want in the OrthoFinder/scripts_of/bin directory and OrthoFinder should look there before looking on the rest of the system.

All the best
David

**Comment by ens-LCampbell** [comment]
Hi David,

Thats the thing, I figured that functionality wasn't working or something as thats exactly what I have tried originally. Placed symlinked to and executibles in the bin dir but it didn't report anything different at all. I figured that it was checking the bin dir. 

I really feel now this isn't an OrthoFinder issue. As I can replicate failing jobs and jobs that do not fail on raxml depending on the compute node its running. I've contact our systems team again to highlight these cases as it should help them pinpoint the issue. 

Best
L
